File: Greedy/798. Smallest Rotation with Highest Score.cpp


/*

Get more point:
    所有数 当从 0 到 A[n-1] 得到一分

Lose point:
    change[i] 表示当rotate 到第i次时, lose 1 point,  
    change[ ( i - A[i] + 1 + N)%N] ++ 比如 N = 6,  i = 3, A[i] = 2,  change[2]++, 表示当rotate step 为2( 2 到index = 1), lose 1 point 
                                       比如 N = 6, i = 2, A[i] = 4,  change[5]++, 表示当rotate step 为5， lose 1 point (4 移到了 index 3)

0 情况特殊, 比如 N = 6, i = 3, A[i] = 0, change[4]++, 表示rotate step 为4 (0 移动到最后一位,), lose 1 point，
    因为所有数当 0 到A[n-1] 都get one point, 0到 A[n-1] lose 1 point, 抵消

*/

class Solution {
public:
    int bestRotation(vector<int>& A) {
        const int N = A.size();
        int change[N] = {0};
        for(int i = 0; i<N; i++)
            change[(i - A[i] + 1 + N)%N]--;
        for(int i = 1; i<N; i++)
            change[i] += change[i-1] + 1;
        return distance(change, max_element(change, change+N));
    }
};


class Solution {
public:
    int bestRotation(vector<int>& A) {
        int N = A.size();
        int counter[N] = {0};
        for (int i = 0; i < N; i++) {
            counter[(i + 1) % N]++;
            counter[(i - (A[i] - 1) + N) % N]--;
            if (i >= A[i]) counter[0]++; //算0的sum
        }
        
        int max = counter[0], bestK = 0;
        for (int i = 1; i < N; i++) {
            counter[i] += counter[i - 1];
            if (counter[i] > max) {
                max = counter[i];
                bestK = i;
            }
        }
        
        return bestK;
    }
};


//2020
class Solution {
public:
    int bestRotation(vector<int>& A) {
        unordered_map<int,int>lose;
        int point = 0, n = A.size();
        for(int i = 0; i<A.size(); ++i){
            if(A[i] <= i) ++point;
            lose[(i-A[i] + n)%n]++;
        }
        int maxRotate = 0, maxPoint = point;
        for(int r = 0; r<n-1; ++r){
            point += 1;
            point -= lose[r];
            if(point > maxPoint){
                maxPoint = point;
                maxRotate = r+1;
            }
        }
        return maxRotate;
    }
};


class Solution {
public:
    int bestRotation(vector<int>& A) {
        unordered_map<int,int>count;
        int n = A.size();
        for(int i = 0; i<A.size(); ++i){
            if(A[i]<=i){
                count[0] += 1; //不用转 得一分
                int j = (i - A[i] + 1) % n;
                count[j] -= 1;
            }
            else{
                int j =  (n - A[i] + i +  1) % n;
                count[j] -= 1;
            }
            count[i+1] += 1; //当这个数移动到最后一位时
        }

        int maxPoint = count[0], point = count[0];
        int bestRotate = 0;
        for(int i = 1; i<n; ++i){
            point += count[i];
            if(point > maxPoint){
                maxPoint = point; 
                bestRotate = i;            
            }
        }
        return bestRotate;
    }
};

File: Greedy/798. Smallest Rotation with Highest Score.py
"""
798. Smallest Rotation with Highest Score

 Given an array A, we may rotate it by a non-negative integer K so that 
 the array becomes A[K], A[K+1], A{K+2], ... A[A.length - 1], A[0], A[1], ..., A[K-1].  
 Afterward, any entries that are less than or equal to their index are worth 1 point. 

For example, if we have [2, 4, 1, 3, 0], and we rotate by K = 2, it becomes [1, 3, 0, 2, 4].  
This is worth 3 points because 1 > 0 [no points], 3 > 1 [no points], 0 <= 2 [one point], 2 <= 3 [one point], 4 <= 4 [one point].

Over all possible rotations, return the rotation index K that corresponds to the highest score we could receive.  
If there are multiple answers, return the smallest such index K.

Example 1:
Input: [2, 3, 1, 4, 0]
Output: 3
Explanation:  
Scores for each K are listed below: 
K = 0,  A = [2,3,1,4,0],    score 2
K = 1,  A = [3,1,4,0,2],    score 3
K = 2,  A = [1,4,0,2,3],    score 3
K = 3,  A = [4,0,2,3,1],    score 4
K = 4,  A = [0,2,3,1,4],    score 3
So we should choose K = 3, which has the highest score.

 

Example 2:
Input: [1, 3, 0, 2, 4]
Output: 0
Explanation:  A will always have 3 points no matter how it shifts.
So we will choose the smallest K, which is 0.

"""
import functools
class Solution:
    def bestRotation(self, A):
        """
        :type A: List[int]
        :rtype: int
        """
        n = len(A)
        delta = [0]*(n+1)
        for i, v in enumerate(A):
            step =  1 + (i - v if i >= v else i + n - v )
            delta[step] += 1
        point = functools.reduce(lambda x, y: x + (1 if y[0]>=y[1] else 0), enumerate(A), 0)
        maxnum = point; minstep = 0
        for i in range(1, n):
            point += 1 - delta[i]
            if point > maxnum:
                point, minstep = maxnum, i
        return minstep

class Solution:
    def bestRotation(self, A):
        N = len(A)
        change = [1] * N
        for i in range(N): change[(i - A[i] + 1) % N] -= 1
        for i in range(1, N): change[i] += change[i - 1]
        return change.index(max(change))